### DISCUSSION TITLE
About the north and south polar voids and layer overlay problems.

### DISCUSSION THREAD

**Discussion by luohuiiqng** [pelicanmapping/osgearth#2408] [Support Q&A] [answered]
When XYZImageLayer was used to load the image map, I observed that the north and south poles were empty. Through checking the data, I learned that most XYZ image resources do not provide the image data of the north and south poles. May I ask how to avoid the appearance of such holes at the code level?
![微信图片_20231221095851](https://github.com/gwaldron/osgearth/assets/82392471/c198bdc0-02ee-4781-8f38-e5b36f884f41)

 When I tried to overlay a high-resolution tiff map of a local area on a lower resolution map, I observed a blank space at the border of the high-resolution map. I tried to solve the problem by setColorBlending, but did not solve it. What should I do to avoid the appearance of blank areas?
![微信图片_20231221094225](https://github.com/gwaldron/osgearth/assets/82392471/7e928a70-4f06-4af7-8920-f698fdcf2829)
Thanks.
```
#include <osgViewer/Viewer>

#include <osgEarth/MapNode>
#include <osgEarth/ImageLayer>
#include <osgEarth/ModelLayer>
#include <osgEarth/GeoTransform>

#include <osgEarth/TMS>
#include <osgEarth/WMS>
#include <osgEarth/GDAL>
#include <osgEarth/XYZ>
#include <osgEarth/Composite>

#include <osgEarth/EarthManipulator>
#include <osgEarth/ExampleResources>
#include <osgEarth/AutoScaleCallback>

#include <osgEarth/TerrainConstraintLayer>
#include <osgEarth/OGRFeatureSource>
#include <osgEarth/LabelNode>

#include <osg/PositionAttitudeTransform>
#include <osgDB/ReadFile>
#include <osgDB/WriteFile>
#include "CustomImageLayer.h"


using namespace osgEarth;
using namespace osgEarth::Util;



/**
 * How to create a simple osgEarth map and display it.
 */
int
main(int argc, char** argv)
{
    osg::setNotifyLevel(osg::NotifySeverity::NOTICE);
    osgEarth::initialize();

    // create the empty map.
    Map* map = new Map();


    XYZImageLayer* osm = new XYZImageLayer();
    //E:/Resources/Gis/yingXiang/{z}/{x}/{y}.jpg
    //E:/Resources/Gis/LPTms/lingP/google/{x}/{y}/{z}.jpg
    //E:/Resources/Gis/LPTms/linpTif/linpTif_Level_16.tif
    //http://[abc].tile.openstreetmap.org/{z}/{x}/{y}.png
    //E:/Resources/Gis/yingXiang/{z}/{x}/{y}.jpg
    osm->setURL("E:/Resources/Gis/yingXiang/{z}/{x}/{y}.jpg");
    osm->setProfile(Profile::create(Profile::SPHERICAL_MERCATOR));
    osm->setCachePolicy(CachePolicy::USAGE_READ_WRITE);
    //osm->setColorBlending(BLEND_INTERPOLATE);   
    osm->setName("XYZ ImageLay");
    map->addLayer(osm);

    

    osg::ref_ptr<GDALImageLayer> terrainLayer = new osgEarth::GDALImageLayer();
    terrainLayer->setURL("E:/Resources/Gis/LPTms/linpTif/linpTif_Level_16.tif");    
    map->addLayer(terrainLayer);



    MapNode* node = new MapNode(map);

    Style labelStyle;
    labelStyle.getOrCreate<TextSymbol>()->alignment() = TextSymbol::ALIGN_CENTER_CENTER;
    labelStyle.getOrCreate<TextSymbol>()->fill()->color() = Color::Red;
    osg::ref_ptr<osgEarth::LabelNode> labelNode = new osgEarth::LabelNode("Ling Ping", labelStyle);
    osgEarth::GeoPoint point(SpatialReference::get("wgs84"), 120.552978515625, 30.30294635121175, 100, osgEarth::ALTMODE_ABSOLUTE);
    labelNode->setPosition(point);
    node->addChild(labelNode);

    osgViewer::Viewer viewer;
    osgEarth::EarthManipulator* emt = new osgEarth::EarthManipulator();
    viewer.setCameraManipulator(emt);
    viewer.getCamera()->setSmallFeatureCullingPixelSize(-1.0f);
    viewer.setSceneData(node);
    //tj jin 117.03  wei 36.05
    Viewpoint vp("home", 120.552978515625, 30.30294635121175, 20000000, 0, -90, 2000);
    emt->setViewpoint(vp);

    return viewer.run();

}
```

**Comment by gwaldron** [comment, marked as answer]
The reason for the "hole" at the poles is because the data is in the Spherical Mercator projection, which only reaches +/- 85 degrees latitude (approximately).

As for the overlay, unless the GeoTIFF has the white color set as the NO-DATA color, it will render. You can use a layer shader to remove it in some cases; here's a simple example:
```xml
<shader>
  <![CDATA[
    #pragma vp_function chroma_key, fragment
    void chroma_key(inout vec4 color) {
        const vec3 no_data = vec3(1,1,1);
        float dist = distance(color.rgb, no_data);
        if (dist <= 0.2)
            color.a = 0.0;
    }
  ]]>
</shader>
```

**Reply by luohuiiqng** [reply]
Ok, maybe I can use a layer with a transparent background to deal with the second problem. Thank you!
Can data in WMST format be loaded using WMSImageLayer? I tried it and it didn't seem to work.

```
#include <osgViewer/Viewer>

#include <osgEarth/MapNode>
#include <osgEarth/ImageLayer>
#include <osgEarth/ModelLayer>
#include <osgEarth/GeoTransform>

#include <osgEarth/TMS>
#include <osgEarth/WMS>
#include <osgEarth/GDAL>
#include <osgEarth/XYZ>
#include <osgEarth/Composite>

#include <osgEarth/EarthManipulator>
#include <osgEarth/ExampleResources>
#include <osgEarth/AutoScaleCallback>
#include <osgEarth/LabelNode>

#include <osgEarth/TerrainConstraintLayer>
#include <osgEarth/OGRFeatureSource>

#include <osg/PositionAttitudeTransform>
#include <osgDB/ReadFile>
#include <osgDB/WriteFile>

using namespace osgEarth;
using namespace osgEarth::Util;

int
usage(int argc, char** argv)
{
    OE_NOTICE
        << "\n" << argv[0]
        << "\n    [--out outFile] : write map node to outFile before exit"
        << std::endl;

    return 0;
}

void
checkErrors(const Layer* layer)
{
    if (layer->getStatus().isError())
    {
        OE_WARN << "Layer " << layer->getName() << " : " << layer->getStatus().message() << std::endl;
    }
}


int
main(int argc, char** argv)
{
    osgEarth::initialize();

    

    // create the empty map.
    Map* map = new Map();
    

    WMSImageLayer* wms = new WMSImageLayer();
    wms->setURL("http://127.0.0.1:9000/bigemap.622dlkte/wmts?access_token=pk.eyJ1IjoiY3VzX2NwdmhlOHJrIiwiYSI6ImNsODFmMDV2bXdxNTByeXdzem91ZDJ2YnciLCJ0IjoxfQ.Y9mLhm18HusLDFz0nE649g10nJNVCKFiV_utj6SN7bE");
    wms->setFormat("png");
    wms->setLayers("hqluo");
    wms->setSRS("EPSG:4326");   
    wms->options().cachePolicy() = CachePolicy::NO_CACHE;
    map->addLayer(wms);
    checkErrors(wms);


    MapNode* node = new MapNode(map);


    Style labelStyle5;
    labelStyle5.getOrCreate<TextSymbol>()->alignment() = TextSymbol::ALIGN_CENTER_CENTER;
    labelStyle5.getOrCreate<TextSymbol>()->fill()->color() = Color::Red;
    osg::ref_ptr<osgEarth::LabelNode> labelNode5 = new osgEarth::LabelNode("HangZhou", labelStyle5);
    osgEarth::GeoPoint point5(SpatialReference::get("wgs84"), 120.1551, 30.2741, 100, osgEarth::ALTMODE_ABSOLUTE);
    labelNode5->setPosition(point5);
    node->addChild(labelNode5);

    // initialize a viewer:
    osgViewer::Viewer viewer;
    EarthManipulator* emt = new EarthManipulator();
    viewer.setCameraManipulator(emt);
    viewer.getCamera()->setSmallFeatureCullingPixelSize(-1.0f);
    viewer.setSceneData(node);
    Viewpoint vp("home", -105.042292, 38.840829, 0, 24.261, -21.6, 3450);
    emt->setViewpoint(vp);
    MapNodeHelper().configureView(&viewer);
  

    return viewer.run();
}



```
Error tips:
void StateSet::setGlobalDefaults() ShaderPipeline disabled.
void StateSet::setGlobalDefaults() ShaderPipeline disabled.
[osgEarth]* XML parsing error
[osgEarth]* Error document empty. (row 0, col 0)
[osgEarth]*  1          unknown/invalid request
[osgEarth]* [WMS] : Service Exception: unknown/invalid request (URI=http://127.0.0.1:9000/bigemap.622dlkte/wmts?access_token=pk.eyJ1IjoiY3VzX2NwdmhlOHJrIiwiYSI6ImNsODFmMDV2bXdxNTByeXdzem91ZDJ2YnciLCJ0IjoxfQ.Y9mLhm18HusLDFz0nE649g10nJNVCKFiV_utj6SN7bE&SERVICE=WMS&VERSION=1.1.1&REQUEST=GetMap&LAYERS=hqluo&FORMAT=image/png&STYLES=&SRS=EPSG:4326&WIDTH=256&HEIGHT=256&BBOX=-180.000000,-90.000000,0.000000,90.000000)
[osgEarth]* XML parsing error
[osgEarth]* Error document empty. (row 0, col 0)
[osgEarth]*  1          unknown/invalid request
[osgEarth]* [WMS] : Service Exception: unknown/invalid request (URI=http://127.0.0.1:9000/bigemap.622dlkte/wmts?access_token=pk.eyJ1IjoiY3VzX2NwdmhlOHJrIiwiYSI6ImNsODFmMDV2bXdxNTByeXdzem91ZDJ2YnciLCJ0IjoxfQ.Y9mLhm18HusLDFz0nE649g10nJNVCKFiV_utj6SN7bE&SERVICE=WMS&VERSION=1.1.1&REQUEST=GetMap&LAYERS=hqluo&FORMAT=image/png&STYLES=&SRS=EPSG:4326&WIDTH=256&HEIGHT=256&BBOX=0.000000,-90.000000,180.000000,90.000000)
[osgEarth]* XML parsing error
[osgEarth]* XML parsing error
[osgEarth]* Error document empty. (row 0, col 0)
)
        unknown/invalid request
[osgEarth]*  [osgEarth]* [WMS] : Service Exception: unknown/invalid request (URI=http://127.0.0.1:9000/bigemap.622dlkte/wmts?access_token=pk.eyJ1IjoiY3VzX2NwdmhlOHJrIiwiYSI6ImNsODFmMDV2bXdxNTByeXdzem91ZDJ2YnciLCJ0IjoxfQ.Y9mLhm18HusLDFz0nE649g10nJNVCKFiV_utj6SN7bE&SERVICE=WMS&VERSION=1.1.1&REQUEST=GetMap&LAYERS=hqluo&FORMAT=image/png&STYLES=&SRS=EPSG:4326&WIDTH=256&HEIGHT=256&BBOX=-180.000000,0.000000,-135.000000,45.000000)
```

```

**Comment by gwaldron** [comment]
(In the future, please post a new question to a new thread - thanks.)

You must use the GDAL WMTS mini-driver to access WMTS in osgEarth. Here is an example:
```xml
<GDALImage name="my wmts layer"> 
    <connection>
        WMTS:https://basemap.nationalmap.gov/arcgis/rest/services/USGSShadedReliefOnly/MapServer/WMTS/1.0.0/WMTSCapabilities.xml
    </connection>
</GDALImage>
```
More information is at https://gdal.org/drivers/raster/wmts.html
Good luck!

**Reply by luohuiiqng** [reply]
Ok, thanks!

**Comment by crystalsky** [comment]
**I use this config for earth file, but the osgearth_imgui couldn't display the map layer, could you give me a help????**
```
<GDALImage name="my wmts layer"> 
    <connection>
        WMTS:https://basemap.nationalmap.gov/arcgis/rest/services/USGSShadedReliefOnly/MapServer/WMTS/1.0.0/WMTSCapabilities.xml
    </connection>
</GDALImage>
```

**Reply by gwaldron** [reply]
Works for me. Maybe you have a network issue?
![image](https://github.com/gwaldron/osgearth/assets/326618/a03ebe90-717b-43ee-98a0-95ea32ddd492)

**Reply by crystalsky** [reply]
Thanks you suggestion, May Be you are right. I will publish the WMTS service by myself on my own computer to test the wmts support.

**Reply by crystalsky** [reply]
Can you provide me with an Earth file format of a WMTS service that your test case can access?

**Reply by gwaldron** [reply]
I just used the one you pasted above

**Reply by crystalsky** [reply]
I successfully configured the WMTS service, but I found that it still couldn't be accessed. Finally, looking at the gdal source code, **it needs to compile the libcurl module. And the runtime path needs the proj.db file.**

**Reply by gwaldron** [reply]
Yes, that makes sense. GDAL in vcpkg includes CURL by default. If you're building GDAL yourself you will need to activate the curl support.

You can also set the `PROJ_LIB` environment variable to help GDAL find proj.db if necessary.
